### ISSUE TITLE
CentOS7.1 Kubelet causes OOM

### ISSUE THREAD

**Issue by Galexrt** [kubernetes/kubernetes#16988]
### Info
* Server is a VM
    * CPU: 2 vCores
    * RAM: 6GB
* OS: `CentOS Linux release 7.1.1503 (Core)`
* Kernel: `Linux node1 4.3.0-1.el7.elrepo.x86_64 #1 SMP Tue Nov 3 20:15:39 EST 2015 x86_64 x86_64 x86_64 GNU/Linux` and the "default" installed `3.10 229` kernel
* Kubernetes version
    * `v1.0.3` runs as expected, no OOMs.
    * `v1.0.7` and `v1.1.1-beta.1` cause OOMs.
* etcd version: `2.2.1`
* flanneld: `0.5.4`
* docker: `1.9.0` (using `docker-engine`, from the https://get.docker.com/ site)
* Running services on the server:
    * glusterfsd
    * etcd
    * flanneld
    * sshd
    * kube-apiserver
    * kube-controller-manager
    * kube-scheduler
    * kubelet
    * docker

### top Output (when OOM occurs)
```
0 13.491g 5.437g      0 S   5.9 93.3   0:21.73 kubelet
```
### free -h (when OOM occurs)
```
Mem:           5.8G        5.6G        105M         16M         82M
        37M
Swap:            0B          0B          0B
```
### sysctl -p Output
```
fs.file-max = 2097152
kernel.panic = 10
net.ipv4.ip_forward = 1
vm.overcommit_memory = 1
vm.overcommit_ratio = 50
vm.panic_on_oom = 0
```
### kubelet.service
```
[Unit]
Description=Kubernetes Kubelet Server
Documentation=https://github.com/kubernetes/kubernetes
After=docker.service
Requires=docker.service

[Service]
WorkingDirectory=/var/lib/kubelet
ExecStart=/usr/bin/kubelet \
	    --v=0 \
	    --api-servers=http://213.136.75.252:8080 \
	    --allow-privileged=true \
	    --address=MY_IP \
	    --port=10250 \
	    --hostname-override=node1 \
	    --enable-load-reader=true \
	    --cluster-domain=cluster.local \
	    --cluster-dns=10.100.0.10 \
	    --configure-cbr0=false \
	    --low-diskspace-threshold-mb=56 \
	    --maximum-dead-containers=50 \
	    --tls-cert-file=/srv/kubernetes/kubelet.crt \
	    --tls-private-key-file=/srv/kubernetes/kubelet.key
Restart=on-failure
RestartSec=5s

[Install]
WantedBy=multi-user.target
```
***
### Description
The OOMs seem to occur whenever I deploy something and/or the cluster does nothing.
I test deployed three busybox pods running the `sleep` command and the system ran into the OOM after only 20 seconds and the kubelet process was using like 13GB of memory.
I looked through the other issues, but it seems like nobody else runs into this problem.

I asked on the `#kubernetes-users` slack, but nobody could help me with this "bug".
For me this seems like a bug, so I'll report it here. I hope you can help me with this bug.

**Comment by Galexrt** [comment]
I still get OOMs on the v1.1.1 version with kubelet.

**Comment by vishh** [comment]
Can you post pprof output from kubelet? You can run `kubectl proxy` and then run the following command:
`go tool pprof http://localhost:8001/api/v1/proxy/nodes/<node_name>/debug/pprof/profile`

**Comment by vishh** [comment]
We haven't tested kubelet against all distributions yet. Thanks for filing and helping debug the issue.

**Comment by Galexrt** [comment]
@vishh Here you go. https://drive.google.com/open?id=0B_zLDa2CD7z4czBjRzVTMkRYSE0
I think the server when I ran the `go tool pprof` on was using v1.1.1-beta.1. With self compiled v1.1.1 binaries I ended up in OOMs too.

**Comment by Galexrt** [comment]
I forgot to activate link sharing for the file. It is available now.

**Comment by vishh** [comment]
Can you post the `png` plot of this data? Once you run `go tool pprof ...`
you can enter `web` to view the png output.

**Comment by Galexrt** [comment]
@vishh Here you go:
* https://drive.google.com/file/d/0B_zLDa2CD7z4d3NwcUVmajVFa3c/view?usp=sharing
* https://drive.google.com/file/d/0B_zLDa2CD7z4dFZjZzFnNlRNNTg/view?usp=sharing
* https://drive.google.com/file/d/0B_zLDa2CD7z4LU94LVV0eWdpaGs/view?usp=sharing
* https://drive.google.com/file/d/0B_zLDa2CD7z4SjIwbWIxdWlWMDA/view?usp=sharing

I looked at the pngs but they all look the same.. The only big difference is that the time is different from picture to picture. Did I do something wrong there?

**Comment by vishh** [comment]
Can you try `pprof` with the kubelet binary? I suspect that symbols might have been stripped out.
`go tool pprof <path-to-kubelet-binary> http://localhost:8001/api/v1/proxy/nodes/<node_name>/debug/pprof/profile`

The kubelet binary will be part of the release bundle you are using, which is usally at `_output/dockerized/bin/linux/amd64/kubelet`.

Apologies for not mentioning this option earlier. I appreciate your help!

**Comment by Galexrt** [comment]
@vishh I added three new files to the folder.
They are named like `pprof-latest-*`.
**I get no more OOMs!**

I really don't know why it is working now..
But it seems like reinstalling the server for the 10th time, installing kubernetes from package and replacing the binaries with the latest release binaries is working..
I think it might was just a temporay issue with the server or the binaries I used.

Thanks for helping me!
